rancher/rancher
# [WIP] Downgrade project monitoring

**Problem:**
Drop the Prometheus from project-level monitoring

**Solution:**
- Modify the app answers when deploying project-level Prometheus
    + Don't clean up any previous answers
    + Introduce a requirement condition to inhibit project-level Prometheus in next version chart
    + Redirect the Grafana to cluster-level Prometheus
- Adjust the logic of the readiness detection of project-level Prometheus
- Drop the useless controller logic
- Modify the app answers when deploying cluster-level Prometheus
    + Change the judgment condition of ServiceMonitor Namespace selector
    + Change the judgment condition for PrometheusRule Namespace selector
- Enable project monitoring to depend on cluster-level

**Related PR:**
https://github.com/rancher/system-charts/pull/115
https://github.com/rancher/types/pull/1019

**Description:**
1. It would not affect the previous running state of the project-level monitoring and cluster-level, after updating the Rancher Server included this PR. (*Rancher don't drop the Prometheus from project-level monitoring and don't allow the cluster-level Prometheus to accept all custom metrics.*)
1. If the users want to drop the Prometheus from the project-level monitoring, they need to upgrade both `project-monitoring` and `cluster-monitoring` apps manually. (*If the users already enabled the project-level monitoring, they should lose the data of project-level Prometheus, but they still could back up the extended dashboards of project-level Grafana. Custom metrics/Alert Rules are irrelevant, after finishing the cluster-monitoring upgrade, cluster-level Prometheus, that only one, would take over all namespaces Custom Metrics/Alert Rules.*)
1. If the users have a requirement for using the project-level monitoring in the future, they still need to upgrade the `cluster-monitoring` app manually.

**Transition:**
- One way to reduce the loss of Custom Metrics/Alert Rules, the user could upgrade the `cluster-monitoring` app to ingest the Custom Metrics or watch the Alert Rules until the *next project-level retention period*.
- However, if the users only enable project-level monitoring, they must enable the cluster-level monitoring to avoid querying errors from project-level Grafana.

**Dependency:**
- In theory, after this PR, project-level monitoring will have a strong dependency relationship with cluster-level, as the display data of project-level Grafana is scraping from cluster-level Prometheus. It means we should never allow project-level monitoring living without cluster-level: 
  - **Could not deploy project-level monitoring without enabling cluster-level**
  - **Disable cluster-level monitoring must drop all project-levels**
- If we implemented both of them, the users would lose their previous project-level Prometheus after upgrading the Rancher. So, I prefer below on this PR:
  - **Could not deploy (new) project-level monitoring without enabling cluster-level**
  - ~**Disable cluster-level monitoring must drop all project-levels**~

**Test plan:**
1. Setup a cluster with the previous version Rancher server that allowed project-level monitoring
1. Enable both cluster-level and project-level monitoring
1. Upgrade the version included this PR
    - [ ] [Verify] the project-level Prometheus should never restart
    - [ ] [Verify] the cluster-level Prometheus should never restart
    - [ ] [Verify] the Custom Metrics could only query from project-level Grafana
    - [ ] [Verify] the Alert Rules could only query from project-level Prometheus
1. Upgrade the cluster-monitoring and project-monitoring apps manually
    - [ ] [Verify] the project-level Prometheus should be dropped
    - [ ] [Verify] the cluster-level Prometheus should never restart
    - [ ] [Verify] the Custom Metrics could query from both project-level and cluster-level Grafana
    - [ ] [Verify] the Alert Rules could query from cluster-level Prometheus
